extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java

```diff
@@ -426,6 +426,16 @@ static <DATA_TYPE> void writeColumnSource(
         LongSupplier rowStepGetter;
         LongSupplier valuesStepGetter;
 
+        // These two lists are parallel, where each element represents a single page. The rawItemCountPerPage
+        // contains the number of items (the sum of the array sizes for each containing row) contained within the
+        // page.
+        final TIntArrayList rawItemCountPerPage = new TIntArrayList();
+
+        // The originalRowsPerPage contains the number of arrays (rows) from the original table contained within
+        // the page.
+        final TIntArrayList originalRowsPerPage = new TIntArrayList();
+        boolean case1 = false;
```

Comment:
This is the just a hack to make the code work.
Mainly the idea is that the variables rowStepGetter and valuesStepGetter are LongSupplier variables.
In case 1 they are initialized with            
            rowStepGetter = originalRowsPerPage.iterator()::next;
            valuesStepGetter = rawItemCountPerPage.iterator()::next;
            (check line 498-499) 
And in other case, they are initialized with a constant 
           rowStepGetter = valuesStepGetter = () -> finalTargetSize;
            (check line 510)
In case 1, they are used to step through the entries in case we have arrays, In step 2, they don't do much.

Now for strings, we have two code paths, we first try to encode them with dictionary and then plain. In case the dictionary encoding fails, we have already stepped through the valuesStepGetter. So we have to reset it before trying to encode plain, otherwise it gives an exception.
So in this hack, I just remembered the initial state and which case it was. But I would really appreciate suggestions on how to implement this better.

Reply:
This was a good find, and the fix is both surgical and suitable. About the only thing I would do differently is try to batch all of this up in a single helper class for "array-encoded" data.